### PR TITLE
Core: Fix `operator[]` for typed dictionaries

### LINKED ISSUE #96772: Typed Dictionaries can allow different typed keys/values when using `operator[]`
### Tested versions

Latest master v4.4.dev.custom_build [21249950d]

### System information

Windows 10

### Issue description

Adding to a typed dictionary using `operator[]` allows adding keys/values of different types, if the types aren't known until run time, or if the typed dictionary is passed into a function that takes an untyped dictionary.  In both cases, using `get_or_add` catches the difference and gives an error.

### Steps to reproduce

Create a new project, and add code below to main scene:

```gdscript
func _ready():
	var dictionary : Dictionary[Node, Node] = { }
	# dictionary["test1"] = 1 # Gives error in editor
	# dictionary.get_or_add(get_key(), get_value()) # Type check error
	dictionary[get_key()] = get_value()
	_test_dictionary_func(dictionary)
	print(dictionary)

func _test_dictionary_func(d : Dictionary):
	# d.get_or_add("test2", 2) # Type check error
	d["test2"] = 2

func get_key():
	return "test1"

func get_value():
	return 1
```

This prints `{ "test1": 1, "test2": 2 }` without giving any errors

### Minimal reproduction project (MRP)

N/A

### PR DESCRIPTION
**Pull request by Repiteo**
- Fixes #96772

Caused by an erroneous regression when rebasing the typed dictionary PR. Reimplements type checking and passing `_p->typed_fallback` to erroneous keys when using `operator[]` on dictionaries.

This is technically a partial fix, because it seems the untyped value assignment via `operator[]` is something that can happen regardless. However, fixing that will probably be a much more involved change, as the passed value isn't linked to the typed dictionary in isolation.

### CODE REVIEW

File: modules/gdscript/tests/scripts/analyzer/errors/typed_dictionary_assign_differently_typed_key.gd
Comment (Repiteo) [reply modules/gdscript/tests/scripts/analyzer/errors/typed_dictionary_assign_differently_typed_key.gd:7]: That indeed suppressed it! Though, the analyzer scenario is still valid, so I'll just use both.

File: modules/gdscript/tests/scripts/runtime/errors/typed_dictionary_assign_differently_typed_value.gd
Comment (dalexeev) [inline modules/gdscript/tests/scripts/runtime/errors/typed_dictionary_assign_differently_typed_value.gd:1]: So, this doesn't completely fix the issue, values ​​can still be assigned with the wrong type:

```gdscript
print(typed) # { 0: "value" }
```

As far as I know we can fix it here (but I'm not sure if this solves the issue for C# and GDExtension):

https://github.com/godotengine/godot/blob/4788f54d9767425bf5435b1cc940885d357795c5/core/variant/variant_setget.cpp#L253-L269

https://github.com/godotengine/godot/blob/4788f54d9767425bf5435b1cc940885d357795c5/core/variant/variant_setget.cpp#L706-L748

Also please avoid multiple error checks in GDScript tests. This is a limitation of the test system, you can only check one error per file, unlike features and warnings.

File: modules/gdscript/tests/scripts/analyzer/errors/typed_dictionary_assign_differently_typed_key.gd
Comment (dalexeev) [reply modules/gdscript/tests/scripts/analyzer/errors/typed_dictionary_assign_differently_typed_key.gd:7]: There is an inconsistency. The following code does not produce a compile time error:

```gdscript
func get_value():
    return "value"

func test():
    var typed: Dictionary[int, int]
    typed[0] = get_value()
    print('not ok')
```

However, the following does

```gdscript
func get_value() -> String: # !!!
    return "value"

func test():
    var typed: Dictionary[int, int]
    typed[0] = get_value()
    print('not ok')
```

This is probably a pre-existing bug. I think we should either fix it now or temporarily remove this test and fix it separately in the future.

Comment (Repiteo) [reply modules/gdscript/tests/scripts/analyzer/errors/typed_dictionary_assign_differently_typed_key.gd:7]: I'll remove the test; that suddenly seems outside the scope of this PR

File: modules/gdscript/tests/scripts/runtime/errors/typed_dictionary_assign_differently_typed_value.gd
Comment (Repiteo) [reply modules/gdscript/tests/scripts/runtime/errors/typed_dictionary_assign_differently_typed_value.gd:1]: Adjusted those two areas to have them handled similar to typed arrays; specifically, passing to a new `set()` function. Unfortunately, it didn't seem to fix the `typed[0] = get_value() # Variant` scenario, so I'll keep that first line as a commented-out section instead.

Comment (dalexeev) [reply modules/gdscript/tests/scripts/runtime/errors/typed_dictionary_assign_differently_typed_value.gd:1]: I missed `VariantKeyedSetGetDictionary`.

```diff
@@ -987,16 +987,11 @@ struct VariantKeyedSetGetDictionary {
 		PtrToArg<Variant>::encode(*ptr, value);
 	}
 	static void set(Variant *base, const Variant *key, const Variant *value, bool *r_valid) {
-		if (VariantGetInternalPtr<Dictionary>::get_ptr(base)->is_read_only()) {
-			*r_valid = false;
-			return;
-		}
-		(*VariantGetInternalPtr<Dictionary>::get_ptr(base))[*key] = *value;
-		*r_valid = true;
+		*r_valid = VariantGetInternalPtr<Dictionary>::get_ptr(base)->set(*key, *value);
 	}
 	static void ptr_set(void *base, const void *key, const void *value) {
 		Dictionary &v = *reinterpret_cast<Dictionary *>(base);
-		v[PtrToArg<Variant>::convert(key)] = PtrToArg<Variant>::convert(value);
+		v.set(PtrToArg<Variant>::convert(key), PtrToArg<Variant>::convert(value));
 	}
 
 	static bool has(const Variant *base, const Variant *key, bool *r_valid) {
```

Comment (Repiteo) [reply modules/gdscript/tests/scripts/runtime/errors/typed_dictionary_assign_differently_typed_value.gd:1]: That did the trick, thanks!

**Review by dalexeev — APPROVED** [review]
Looks good to me, but it would be nice to get a second review.

File: core/variant/dictionary.cpp
Comment (Repiteo) [reply core/variant/dictionary.cpp:87]: I'm personally a big fan of the "proper" syntax implemented in `gdextension_interface.h`, though that only works in a docstrings & is primarily suited for exposed/binding contexts. I'm down for a universal `WARNING: ` preface, akin to `TODO: ` & `FIXME: `, though that'll need to be specified [in the documentation](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

**Review by akien-mga — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
